wit/outgoing-handler.wit

```diff
@@ -0,0 +1,15 @@
+// The `wasi:http/outgoing-handler` interface is meant to be imported by
+// components and implemented by the host.
+//
+//   NOTE: in Preview3, this interface will be merged with
+//   `wasi:http/outgoing-handler` into a single `wasi:http/handler` interface
+//   that takes a `request` parameter and returns a `response` result.
+//
+default interface outgoing-handler {
+  use pkg.types.{outgoing-request, incoming-response, error}
+
+  // The parameter and result types of the `handle` function allow the caller
+  // to concurrently stream the bodies of the outgoing request and the incoming
+  // response.
+  handle: func(request: outgoing-request) -> result<incoming-response, error>
```

Comment:
Yep!  That does raise a really good point about something I may have overlooked here: for Preview2, should this function return a resource that can be passed to `poll_oneoff` (just like the streams), thereby allowing concurrent execution before the `result` is known?  IIUC, knowing whether to return the success or failure case requires a network round trip which may result in a timeout so it can take a "long" time, so it seems like "yes", we do need to emulate a `future` here.  Is that right, though?

Reply:
Yes, I would believe so. To me it seems that we need to take this into account because there is at least one scenario that a `future` applies.

Reply:
I guess we could get away with not worrying about this as long as the different languages' codegen can convert an implementation that returns `future` to one that does not internally. In very practical terms, I would expect the `handle` function to have the following signatures in these languages:
- **Rust**: `async fn handle(request: OutgoingRequest) -> Result<IncomingResponse, HttpError> {}` (similar to code [here](https://github.com/bytecodealliance/preview2-prototyping/blob/b2382e692bbee7ce189dd23777e76e45abd723b3/host/src/http.rs#L7-L33))
- **TypeScript**: `async function handle(request: OutgoingRequest) -> Promise<Monad<IncomingResponse, HttpError>> {}` (similar to code [here](https://github.com/bytecodealliance/jco/blob/2c5c3bc6bc3ad49f0da9d7ec9e2a44e084432a28/packages/preview2-shim/lib/http/make-request.js))

Reply:
Ok, I'll add a pseudo-future resource (for the pre-`future` Preview2 timeframe) to use in the return type of `outgoing-handler`.  With native `future`/`stream` types, it's easy enough to see how a bindings generator can know to produce `Promise`s and `ReadableStream`s.  But in the Preview2 timeframe, when it's all just opaque resource types (from the binding generator's POV), this adaptation may need to happen manually in handwritten glue code.

Reply:
Thanks for the explanation. I think I understand better now the difficulties.

Reply:
If you believe waiting a bit will allow the wit-bindgen tooling to mature the partial async support (at least to a point where a future will not cause an error to be thrown). We can leave this piece to a next iteration of this WIT definition.

Reply:
Also, we should probably have a couple timeouts built in here. Specifically, an overall timeout and a connect timeout. I believe they should be separate because they will typically have very different sizes, i.e. 1 second vs. 30 seconds, and the correct response to them will differ. For instance, a connect timeout can be safely retried, whereas a normal timeout on a POST request cannot be. 

There are other timeouts we might consider, which each have different uses, such as lookup, time-to-first-byte, time-to-last-byte, and time-between-bytes.

Reply:
Updated with pseudo-future [here](https://github.com/WebAssembly/wasi-http/pull/3/commits/eaa35579fd9c7efb5ca0618eb61ed555548662f8) (and updated wasm-io deps).

Reply:
Added optional timeout parameters [here](https://github.com/WebAssembly/wasi-http/pull/3/commits/2118d730d61ad6e963b9eeb2d173d505a2c1d551).  Initially, I included the timeout options in the `incoming-handler` interface, but because you can't directly chain components pre-Preview3, and thus `incoming-handler` is primarily going to be called by the host, not another component, I couldn't really see how it would be used so I [removed](https://github.com/WebAssembly/wasi-http/pull/3/commits/ec890b2021a6663283b2a83dda3158ce1e421bde) it, but happy to add it back if there is a use.

Reply:
Just wanted to mention we've had recent requests for concurrent outbound HTTP support in Spin, so we're really excited about your pseudo-future update, @lukewagner!